### PR TITLE
Abstract away service name

### LINKED ISSUE #259: [FEATURE] Add support for OpenSearch Serverless
### Is your feature request related to a problem?

- OpenSearch serverless service identifier changes from `es` to `aoss`
- Sigv4 requires a hash

### What solution would you like?

Document support for OpenSearch Serverless in the compatibility matrix, add integration tests.

### Do you have any additional context?

Coming from https://github.com/opensearch-project/opensearch-clients/issues/44

### PR DESCRIPTION
**Pull request by harshavamsi**
Signed-off-by: Harsha Vamsi Kalluri <harshavamsi096@gmail.com>

### Description
Removes hard coded service name. Users can specify their own service name while signing requests for SigV4.

### Issues Resolved
Closes #259 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -26,6 +26,7 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - Updated getting started to user guide ([#233](https://github.com/opensearch-project/opensearch-py/pull/233))
 - Updated CA certificate handling to check OpenSSL environment variables before defaulting to certifi ([#196](https://github.com/opensearch-project/opensearch-py/pull/196))
 - Updates `master` to `cluster_manager` to be inclusive ([#242](https://github.com/opensearch-project/opensearch-py/pull/242))
+- Updates the service name in SigV4 signing to allow users to specify custom input ([#268](https://github.com/opensearch-project/opensearch-py/pull/268))
```
Comment (dblock) [inline CHANGELOG.md:29]: "Support a custom signing service name for AWS SigV4"

**Review by dblock — COMMENTED** [review]
Needs tests, otherwise LGTM.

File: CHANGELOG.md (hunk 1 above)
Comment (harshavamsi) [reply CHANGELOG.md:29]: Thanks, yeah this is not ready yet. Will need to remove ContentLength Headers.

File: opensearchpy/helpers/asyncsigner.py [hunk 2]
```diff
@@ -37,18 +41,52 @@ def _sign_request(self, method, url, query_string, body):
...
+            if hasattr(aws_request.body, "read"):
+                aws_request.body = aws_request.body.read()
+            self.encode_body(aws_request)
+            content_hash = hashlib.sha256(aws_request.body)
+        elif hasattr(aws_request, "content") and aws_request.content is not None:
+            content_hash = hashlib.sha256(aws_request.content)
+        else:
+            content_hash = hashlib.sha256(b"")
+
+        aws_request.headers["x-amz-content-sha256"] = content_hash.hexdigest()
```
Comment (dblock) [inline opensearchpy/helpers/asyncsigner.py:66]: AFAIK this is not necessary for either managed or serverless.

File: opensearchpy/helpers/asyncsigner.py [hunk 3]
```diff
@@ -37,18 +41,52 @@ def _sign_request(self, method, url, query_string, body):
...
+
+        sig_v4_auth = SigV4Auth(self.credentials, self.service, self.region)
         sig_v4_auth.add_auth(aws_request)
 
+        if "Content-Length" in aws_request.headers:
```
Comment (dblock) [inline opensearchpy/helpers/asyncsigner.py:71]: You would need to delete the content-length header before signing.

**Review by dblock — CHANGES_REQUESTED** [review]
The change of deleting `content-length` seems hidden in here. I believe this is to support Serverless and managed alike. I believe this won't be necessary when serverless GAs, so maybe remove it from this PR and open another specifically for the serverless workaround?

File: USER_GUIDE.md [hunk 4]
```diff
@@ -407,8 +407,9 @@ import boto3
 
 host = '' # cluster endpoint, for example: my-test-domain.us-east-1.es.amazonaws.com
 region = 'us-west-2'
+service = 'es' # 'aoss' for OpenSearch serverless
```
Comment (dblock) [inline USER_GUIDE.md:410]: serverless -> Serverless

File: USER_GUIDE.md [hunk 5]
```diff
@@ -450,8 +451,9 @@ import boto3
...
```
Comment (dblock) [inline USER_GUIDE.md:454]: serverless -> Serverless

**Review by dblock — COMMENTED** [review]
One more nit, sorry :)

File: opensearchpy/helpers/asyncsigner.py [hunk 6]
```diff
@@ -28,6 +26,10 @@ def __init__(self, credentials, region):  # type: ignore
             raise ValueError("Region cannot be empty")
         self.region = region
 
+        if not region:
```
Comment (dblock) [inline opensearchpy/helpers/asyncsigner.py:29]: Did you mean `if not service`?

Comment (harshavamsi) [reply opensearchpy/helpers/asyncsigner.py]: nice catch. added tests.

**Review by dblock — APPROVED** [review]

**Pull request merged**